internal/buffer/unbounded.go

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package buffer provides an implementation of an unbounded buffer.
+package buffer
+
+import "sync"
+
+// Unbounded is an implementation of an unbounded buffer which does not use
+// extra goroutines. This is typically used for passing updates from one entity
+// to another within gRPC.
+//
+// All methods on this type are thread-safe and don't block on anything except
+// the underlying mutex used for synchronization.
+type Unbounded struct {
+	c       chan interface{}
+	mu      sync.Mutex
+	backlog []interface{}
+}
+
+// NewUnbounded returns a new instance of Unbounded.
+func NewUnbounded() *Unbounded {
+	return &Unbounded{c: make(chan interface{}, 1)}
+}
+
+// Put adds t to the unbounded buffer.
+func (b *Unbounded) Put(t interface{}) {
+	b.mu.Lock()
+	if len(b.backlog) == 0 {
+		select {
+		case b.c <- t:
+			b.mu.Unlock()
+			return
+		default:
+		}
+	}
+	b.backlog = append(b.backlog, t)
+	b.mu.Unlock()
+}
+
+// Load sends the earliest buffered data, if any, onto the read channel
+// returned by Get(). Users are expected to call this every time they read a
+// value from the read channel.
+func (b *Unbounded) Load() {
+	b.mu.Lock()
+	if len(b.backlog) > 0 {
+		select {
+		case b.c <- b.backlog[0]:
+			b.backlog[0] = nil
+			b.backlog = b.backlog[1:]
+		default:
+		}
+	}
+	b.mu.Unlock()
+}
+
+// Get returns a read channel on which values added to the buffer, via Put(),
+// are sent on.
+//
+// Upon reading a value from this channel, users are expected to call Load() to
+// send the next buffered value onto the channel if there is any.
+func (b *Unbounded) Get() <-chan interface{} {
+	return b.c
```

Comment:
Is the buffer ordered? If Put(A) happens before Put(C), is Get guaranteed to return A before C? If so, consider this series of events where goroutines 1 and 2 are executing concurrently.

Goroutine 1:

1.1. Put(A)
1.2. Put(B)
1.3. Get()
1.4. Load()

Goroutine 2:

2.1. Put(C)

Between 1.3 and 1.4, let's say 2.1 happens. Since `<- b.Get` has completed, the chan is empty, so 2.1 puts C into the chan. Now the channel has C, but the backlog has B, so C would be returned before B the next time Get() is called even though B was inserted before C. (1.4's execution doesn't matter anymore because Load's select executes the empty default and simply returns.)

If this isn't ordered, ignore everything I said :D (but maybe add a note on the Get saying that this may return values out of order, though?)

P.S. I know that this is just code being moved, but thought it'd be interesting to mention.

Reply:
`Get` doesn't actually return values. It simply returns the channel and it is the caller's responsibility to read from that channel. So, in your example when `1.3` returns, it does not mean that the underlying channel in the buffer implementation has been drained. It just means that `goroutine-1` has a channel from which it can read the same two values that it pushed.

Reply:
I understand that, but to clarify, I meant `x <- b.Get()` when I said "1.3. Get()":

```go
barrier := make(chan bool)

go func() {
  b.Put(1)
  b.Put(2)
  v1 <- b.Get()
  barrier <- true
  b.Load()
  v2 <- b.Get()
  b.Load()
}()

<- barrier // used to ensure that b.Put(3) strictly happens after b.Put(2)

go func() {
  b.Put(3)
  v3 <- b.Get()
  b.Load()
}()
```

The question is: is `v2` guaranteed to be 2?

I think `v2 = 3`, `v3 = 2` is possible if `b.Put(3)` happens before the post-barrier `b.Load()`, which breaks ordering, no?

Reply:
You are right with your last statement, but I don't think that breaks ordering. This buffer has the same behavior as a buffered channel. In your example, if you used a vanilla channel with a buffer of 3, you can still run into the same issue that you have mentioned here.

Reply:
True. I assumed that since `Put` is non-blocking (unlike a vanilla channel), a user might expect the result to be in the order that `Put`s returned (like a vanilla channel). That is, in a vanilla channel of capacity 2, if `ch <- x` completes before `ch <- y`, then receiving from the channel guarantees that `x` will be returned before `y`. This buffer, however, does not _guarantee_ it, correct?

Reply:
Note that I impose no restriction on which channel send completes first (that's entirely random); just that if `x` completes first, then `x` will be returned first when receiving from the channel.

Reply:
Coming back to your original question:

```
The question is: is v2 guaranteed to be 2?
```

If you can guarantee that the assignment to `v2` happens before the assignment to `v3`, then we can guarantee that `v2 will be equal to 2`.

If you look at it from the point of view of reading from the channel returned from `Get`, it still maintains order. Now, if you use the returned channel in different goroutines, the order in which you see values will be the order in which the reads execute. 

Maybe I'm missing your point.